RapidReact/src/main/java/frc/robot/subsystem/ClimberSubsystem.java

```diff
@@ -1,112 +1,311 @@
-  private final Loggable<String> hookToggleState = BucketLog.loggable(Put.STRING, "climber/hookState");
-  private final Loggable<String> elevatorToggleState = BucketLog.loggable(Put.STRING, "climber/elevatorState");
+  private final Loggable<String> elevatorTiltedState = BucketLog.loggable(Put.STRING, "climber/elevatorTiltedState");
+  private final Loggable<Double> climberMotorPosition = BucketLog.loggable(Put.DOUBLE, "climber/climberMotorPosition");
 
   public ClimberSubsystem(Config config) {
     super(config);
   }
 
   @Override
   public void init() {
-    climber = new WPI_TalonSRX(config.climberMotor_ID);
+    climber1 = MotorUtils.makeSRX(config.climber.climber1);
+    climber2 = MotorUtils.makeSRX(config.climber.climber2);
+
     if (config.enablePneumatics) {
-      elevatorSolenoid = new DoubleSolenoid(PneumaticsModuleType.REVPH, config.
-      elevatorSolenoid_ID1, config.elevatorSolenoid_ID2);
-      fixedHookSolenoid = new DoubleSolenoid(PneumaticsModuleType.REVPH, config.fixedHookSolenoid_ID1, config.fixedHookSolenoid_ID2);
+      elevatorSolenoid =
+        new DoubleSolenoid(PneumaticsModuleType.REVPH, config.elevatorSolenoid_ID1, config.elevatorSolenoid_ID2);
     }
+
+    currentClimbState = ClimbState.Idle;
+  }
+
+  boolean isClimberAtSetpoint()
+  {
+    climbErrThreshold1 = 10;
+    climbLoopsToSettle1 = 10;
+    withinThresholdLoops1 = 0;
+    boolean climber1AtSetpoint = false;
+    if (climber1.getClosedLoopError() < +climbErrThreshold1 && climber1.getClosedLoopError() > -climbErrThreshold1) {
+      withinThresholdLoops1++;
+
+      if (withinThresholdLoops1 > climbLoopsToSettle1)
+        climber1AtSetpoint = true;
+    } else {
+      withinThresholdLoops1 = 0;
+    }
+    
+    climbErrThreshold2 = 10;
+    climbLoopsToSettle2 = 10;
+    withinThresholdLoops2 = 0;
+    boolean climber2AtSetpoint = false;
+    if (climber2.getClosedLoopError() < +climbErrThreshold2 && climber2.getClosedLoopError() > -climbErrThreshold2) {
+      withinThresholdLoops2++;
+
+      if (withinThresholdLoops2 > climbLoopsToSettle2)
+        climber2AtSetpoint = true;
+    } else {
+      withinThresholdLoops2 = 0;
+    }
+
+    return climber1AtSetpoint && climber2AtSetpoint;
+  }
+
+  private void idleInit()
+  {
+    setElevatorTilted(false);
+    autoRetractFull();
+  }
+
+  private void extendPartialInit()
+  {
+    setElevatorTilted(false);
+    autoExtendPartial();
+  }
+
+  private void extendFullInit()
+  {
+    setElevatorTilted(true);
+    autoExtendFull();
+  }
+
+  private void retractInit()
+  {
+    setElevatorTilted(false);
+    autoRetractFull();
   }
 
   @Override
-  public void periodic() {}
+  public void periodic() {
+    if (!autoClimb)
+      return;
+    
+    ClimbState nextState = ClimbState.Idle;
+
+    switch (currentClimbState)
+    {
+      case Idle:
+        if (autoClimbPressed)
+        {
+          nextState = ClimbState.ExtendPartial;
+          extendPartialInit();
```

Comment:
This looks pretty nice and should work fine as-is. If you wanted, can clean up the logic a little more and avoid repetition by having a separate switch statement for the actions and calling the relevant functions each loop. There's no harm in calling extendFullInit, for example, every loop, not just when the state is entered.

Reply:
Heh, I told @ojas-sanghi yesterday it was cleaner not to call the set() every loop. :)

Reply:
Haha, I like to think of “I’m in this state, so I do this thing”, not “I have this state transition, so I do this thing”. But it doesn’t really matter as long as you’re consistent and understand the abstraction.